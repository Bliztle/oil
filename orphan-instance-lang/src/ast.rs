use lalrpop_util::lalrpop_mod;

lalrpop_mod!(
    #[allow(clippy::ptr_arg)]
    #[rustfmt::skip]
    pub oil
);

#[derive(Debug)]
pub struct Program {
    pub nodes: Vec<Definition>,
}

#[derive(Debug)]
pub struct Ident(String);

#[derive(Debug)]
pub enum Definition {
    Struct(Ident, Vec<StructField>),
    Trait(Ident, Vec<TraitMethod>),
    Impl(Impl),
    Mod,
    Func(Function),
}

#[derive(Debug)]
pub struct StructField {
    ident: Ident,
    field_type: Type,
}

#[derive(Debug)]
pub struct TraitMethod {
    ident: Ident,
    parameters: Vec<Parameter>,
    return_type: Type,
}

#[derive(Debug)]
pub struct Parameter {
    ident: Ident,
    parameter_type: Type,
}

#[derive(Debug)]
pub enum Expr {
    Literal(Literal),
    Ident(Ident),
    Paren(Box<Expr>),
    Binary(Box<Expr>, BinOp, Box<Expr>),
    Let(Ident, Box<Expr>),
    If(Box<Expr>, Box<Expr>, Box<Expr>),
    Block(Vec<Expr>),
    Invocation(Ident, Vec<Expr>),
}

#[derive(Debug)]
pub enum Literal {
    I32(i32),
    F32(f32),
}

#[derive(Debug)]
pub enum BinOp {
    Mult,
    Div,
    Plus,
    Minus,
    LT,
    LTE,
    GT,
    GTE,
    Equal,
}

#[derive(Debug)]
pub enum Type {
    Unit,
    I32,
    U32,
    Function(Vec<Parameter>, Box<Type>),
    Ident(Ident),
}

#[derive(Debug)]
pub enum Impl {
    /** Impl name, Trait name, Struct name */
    Trait(Option<Ident>, Ident, Ident, Vec<Function>),
    Struct(Ident, Vec<Function>),
}

#[derive(Debug)]
pub struct Function {
    ident: Ident,
    parameters: Vec<Parameter>,
    return_type: Option<Type>,
    body: Expr,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_func() {
        let parser = oil::funcParser::new();

        let good = [
            "fn my_function () {}",
            "fn my_function ( ) { }",
            "fn my_function (a: A) { a }",
            "fn my_function (a: A) -> A { a }",
        ];

        let bad = [
            "fn my_function my_function () {}",
            "fn my_function ( } {)",
            "fn my_function (a: A) { fn }",
            "fn my_function (a A) { }",
            "fn my_function (a) { }",
        ];

        assert!(good.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_ok()));
        assert!(bad.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_err()));
    }

    #[test]
    fn test_parse_impl() {
        let parser = oil::implParser::new();
        let func = "fn my_func() {}";

        let good = [
            format!("impl MyStruct {{ {func} }}"),
            format!("impl MyImpl of MyTrait for MyStruct {{ {func} }}"),
            format!("impl MyTrait for MyStruct {{ {func} }}"),
        ];

        let bad = [
            format!("impl {{ {func} }}"),
            format!("impl MyStruct MyStruct {{ {func} }}"),
            format!("impl MyImpl of for MyStruct {{ {func} }}"),
            format!("impl MyImpl MyTrait for MyStruct {{ {func} }}"),
        ];

        assert!(good.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_ok()));
        assert!(bad.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_err()));
    }

    #[test]
    fn test_parse_trait() {
        let parser = oil::traitParser::new();

        let good = [
            "trait MyTrait {}",
            "trait MyTrait { }",
            "trait MyTrait { fn my_func() -> A }",
            "trait MyTrait { fn my_func(a: A) -> A }",
            "trait MyTrait { fn my_func(a: A) -> () }",
        ];

        let bad = ["trait trait", "trait MyTrait"];

        assert!(good.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_ok()));
        assert!(bad.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_err()));
    }

    #[test]
    fn test_parse_ident() {
        let parser = oil::identParser::new();

        let good = ["a1", "_a", "_", "kjdngf_34", "gdfg3"];

        let bad = [
            "1a", "sdf@f", // Keywords
            "fn", "trait", "impl", "for", "mod", // "if",
            // "else",
            "i32", "u32",
        ];

        assert!(good.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_ok()));
        assert!(bad.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_err()));
    }

    #[test]
    fn test_parse_expr() {
        let parser = oil::exprParser::new();

        let good = [
            "1",
            "324908",
            "0.654",
            "234.9",
            "a",
            "fdsf3",
            "1 + 2",
            "{ 1 }",
            "( 1 )",
            "( 1 + 2 )",
            "if 1 { 1 } else { 2 }",
            "if (3) { 1 } else { 2 }",
            "if (3 == 2) { 1 } else { 2 }",
            "if 3 > 2 { 1 } else { 2 }",
        ];

        let bad = ["543.", "4s", "let 2 = 3", "1 2", "1,2", "{ 1 2 }"];

        assert!(good.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_ok()));
        assert!(bad.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_err()));
    }

    #[test]
    fn test_parse_expr_statement() {
        let parser = oil::expr_statementParser::new();

        let good = [
            "if 1 { 1 } else { 2 }",
            "if (3) { 1 } else { 2 }",
            "if (3 == 2) { 1 } else { 2 }",
            "if 3 > 2 { 1 } else { 2 }",
            "let x = 3",
        ];

        let bad = [
            "1",
            "324908",
            "0.654",
            "234.9",
            "a",
            "fdsf3",
            "1 + 2",
            "( 1 )",
            "( 1 + 2 )",
        ];

        assert!(good.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_ok()));
        assert!(bad.iter().all(|s| dbg!(parser.parse(dbg!(s))).is_err()));
    }
}
