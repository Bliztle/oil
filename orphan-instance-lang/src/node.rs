//! This module defines the Node wrapper used in the AST
//! It exposes an annotation system which strictly enforces type safety on existing annotations,
//! only allowing get / set operations with preconfigured types

use crate::ast::Type;
use std::{
    any::{Any, TypeId},
    collections::HashMap,
    marker::PhantomData,
};
use typelist::typelist;

pub trait Annotation: 'static {}

// struct Type;
typelist!(1, Type);

impl Annotation for Type {}

#[derive(Debug)]
pub struct Node<T, S = Nil> {
    pub node: T,
    annotations: HashMap<TypeId, Box<dyn Any>>,
    _state: PhantomData<S>,
}

impl<T> From<T> for Node<T> {
    fn from(value: T) -> Self {
        Node::new(value)
    }
}

impl<T> From<T> for Box<Node<T>> {
    fn from(value: T) -> Self {
        Node::from(value).into()
    }
}

impl<T, S> Node<T, S> {
    pub fn new(value: T) -> Self {
        Node {
            node: value,
            annotations: HashMap::new(),
            _state: PhantomData,
        }
    }

    /// Adds the provided annotation to the node, and returns a new node with the updated signature
    /// This invalidates the old node, as it is impossible to retrieve a value not included in the
    /// node's signature. Only the new node should be used from then on.
    pub fn add_annotation<U>(mut self, annotation: U) -> Node<T, Cons<U, S>>
    where
        U: Annotation,
        S: Excludes<U>,
    {
        self.annotations
            .insert(TypeId::of::<U>(), Box::new(annotation));
        Node {
            node: self.node,
            annotations: self.annotations,
            _state: PhantomData,
        }
    }

    pub fn add_annotation_unchecked<U>(&mut self, annotation: U)
    where
        U: Annotation,
        S: Excludes<U>,
    {
        self.annotations
            .insert(TypeId::of::<U>(), Box::new(annotation));
    }

    pub fn change_annotation<U>(&mut self, annotation: U)
    where
        U: Annotation,
        S: Includes<U>,
    {
        self.annotations
            .insert(TypeId::of::<U>(), Box::new(annotation));
    }

    /// # Panics
    pub fn get_annotation<U>(&self) -> &U
    where
        U: Annotation,
        S: Includes<U>,
    {
        let annotation = self
            .annotations
            .get(&TypeId::of::<U>())
            .and_then(|boxed| boxed.downcast_ref::<U>());
        annotation
            .expect("Annotation did not exist, which should have been guaranteed by type system")
    }

    pub fn get_annotation_unchecked<U>(&self) -> Option<&U>
    where
        U: Annotation,
        S: Excludes<U>,
    {
        self.annotations
            .get(&TypeId::of::<U>())
            .and_then(|boxed| boxed.downcast_ref::<U>())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_add_get_annotation() {
        let node: Node<i32> = 0.into();
        let mut node = node.add_annotation(Type::I32);
        let _: &Type = node.get_annotation();
        node.change_annotation(Type::F32);
    }
}
