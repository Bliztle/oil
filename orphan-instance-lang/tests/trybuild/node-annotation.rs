extern crate orphan_instance_lang;
use orphan_instance_lang::ast::Type;
use orphan_instance_lang::node::Node;

#[rustfmt::skip]
fn doesnt_compile() {
    let node: Node<i32> = 0.into();
    let _: &Type = node.get_annotation(); // Shouldn't be able to get annotation not added
    node.change_annotation(Type::I32); // Shouldn't be able to change annotation not added
    let node = node.add_annotation(Type::I32);
    let node = node.add_annotation(Type::I32); // Shouldn't be able to add annotation twice
    let _ = node.add_annotation_unchecked(Type::I32); // Shouldn't be able to add unchecked when annotation is already checked
    let _: &Type = node.get_annotation_unchecked(); // Shouldn't be able to get unchecked when annotation is already checked
}

fn main() {}
