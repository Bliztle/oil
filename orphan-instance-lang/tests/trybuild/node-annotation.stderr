error[E0277]: the trait bound `Nil: Annotation` is not satisfied
 --> tests/trybuild/node-annotation.rs:8:25
  |
8 |     let _: &Type = node.get_annotation(); // Shouldn't be able to get annotation not added
  |                         ^^^^^^^^^^^^^^ the trait `Annotation` is not implemented for `Nil`
  |
  = help: the trait `Annotation` is implemented for `Type`
note: required by a bound in `Node::<T, S>::get_annotation`
 --> src/node.rs
  |
  |     pub fn get_annotation<U>(&self) -> &U
  |            -------------- required by a bound in this associated function
  |     where
  |         U: Annotation,
  |            ^^^^^^^^^^ required by this bound in `Node::<T, S>::get_annotation`

error[E0308]: mismatched types
 --> tests/trybuild/node-annotation.rs:8:20
  |
8 |     let _: &Type = node.get_annotation(); // Shouldn't be able to get annotation not added
  |            -----   ^^^^^^^^^^^^^^^^^^^^^ expected `&Type`, found `&Nil`
  |            |
  |            expected due to this
  |
  = note: expected reference `&Type`
             found reference `&Nil`

error[E0277]: the trait bound `Nil: Annotation` is not satisfied
 --> tests/trybuild/node-annotation.rs:9:28
  |
9 |     node.change_annotation(Type::I32); // Shouldn't be able to change annotation not added
  |          ----------------- ^^^^^^^^^ the trait `Annotation` is not implemented for `Nil`
  |          |
  |          required by a bound introduced by this call
  |
  = help: the trait `Annotation` is implemented for `Type`
note: required by a bound in `Node::<T, S>::change_annotation`
 --> src/node.rs
  |
  |     pub fn change_annotation<U>(&mut self, annotation: U)
  |            ----------------- required by a bound in this associated function
  |     where
  |         U: Annotation,
  |            ^^^^^^^^^^ required by this bound in `Node::<T, S>::change_annotation`

error[E0308]: mismatched types
 --> tests/trybuild/node-annotation.rs:9:28
  |
9 |     node.change_annotation(Type::I32); // Shouldn't be able to change annotation not added
  |          ----------------- ^^^^^^^^^ expected `Nil`, found `Type`
  |          |
  |          arguments to this method are incorrect
  |
note: method defined here
 --> src/node.rs
  |
  |     pub fn change_annotation<U>(&mut self, annotation: U)
  |            ^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `Cons<Type, Nil>: Excludes<_>` is not satisfied
  --> tests/trybuild/node-annotation.rs:11:36
   |
11 |     let node = node.add_annotation(Type::I32); // Shouldn't be able to add annotation twice
   |                     -------------- ^^^^^^^^^ the trait `Excludes<_>` is not implemented for `Cons<Type, Nil>`
   |                     |
   |                     required by a bound introduced by this call
   |
   = help: the trait `Excludes<T>` is implemented for `Nil`
note: required by a bound in `Node::<T, S>::add_annotation`
  --> src/node.rs
   |
   |     pub fn add_annotation<U>(mut self, annotation: U) -> Node<T, Cons<U, S>>
   |            -------------- required by a bound in this associated function
...
   |         S: Excludes<U>,
   |            ^^^^^^^^^^^ required by this bound in `Node::<T, S>::add_annotation`

error[E0277]: the trait bound `Cons<Type, Cons<Type, Nil>>: Excludes<_>` is not satisfied
  --> tests/trybuild/node-annotation.rs:12:43
   |
12 |     let _ = node.add_annotation_unchecked(Type::I32); // Shouldn't be able to add unchecked when annotation is already checked
   |                  ------------------------ ^^^^^^^^^ the trait `Excludes<_>` is not implemented for `Cons<Type, Cons<Type, Nil>>`
   |                  |
   |                  required by a bound introduced by this call
   |
   = help: the trait `Excludes<T>` is implemented for `Nil`
note: required by a bound in `Node::<T, S>::add_annotation_unchecked`
  --> src/node.rs
   |
   |     pub fn add_annotation_unchecked<U>(&mut self, annotation: U)
   |            ------------------------ required by a bound in this associated function
...
   |         S: Excludes<U>,
   |            ^^^^^^^^^^^ required by this bound in `Node::<T, S>::add_annotation_unchecked`

error[E0277]: the trait bound `Cons<Type, Cons<Type, Nil>>: Excludes<_>` is not satisfied
  --> tests/trybuild/node-annotation.rs:13:25
   |
13 |     let _: &Type = node.get_annotation_unchecked(); // Shouldn't be able to get unchecked when annotation is already checked
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Excludes<_>` is not implemented for `Cons<Type, Cons<Type, Nil>>`
   |
   = help: the trait `Excludes<T>` is implemented for `Nil`
note: required by a bound in `Node::<T, S>::get_annotation_unchecked`
  --> src/node.rs
   |
   |     pub fn get_annotation_unchecked<U>(&self) -> Option<&U>
   |            ------------------------ required by a bound in this associated function
...
   |         S: Excludes<U>,
   |            ^^^^^^^^^^^ required by this bound in `Node::<T, S>::get_annotation_unchecked`

error[E0308]: mismatched types
  --> tests/trybuild/node-annotation.rs:13:20
   |
13 |     let _: &Type = node.get_annotation_unchecked(); // Shouldn't be able to get unchecked when annotation is already checked
   |            -----   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&Type`, found `Option<&_>`
   |            |
   |            expected due to this
   |
   = note: expected reference `&Type`
                   found enum `Option<&_>`
help: consider using `Option::expect` to unwrap the `Option<&_>` value, panicking if the value is an `Option::None`
   |
13 |     let _: &Type = node.get_annotation_unchecked().expect("REASON"); // Shouldn't be able to get unchecked when annotation is already checked
   |                                                   +++++++++++++++++
